### PR TITLE
k3s: update script fixed

### LINKED ISSUE #156615: k3s: update script is broken
Due to changes to K3s's download script update script is currently broken.

Currently I plan fixing the update script after #158089 is merged.

### PR DESCRIPTION
**Pull request by superherointj**
k3s: update script fixed

Fixes #156615

## Changes

* Several fixes to update script for supporting Go Modules.

* Remove duplication of (I guess this is desirable?):
```
    "-X github.com/containerd/containerd/version.Version=${containerdVersion}"
    "-X github.com/containerd/containerd/version.Package=github.com/k3s-io/containerd"
```
  * Standardize on versions not having 'v', as Nix packages version are standardized to not have 'v'.

@euank

--

Check for logs at:
https://r.ryantm.com/log/updatescript/k3s/

### CODE REVIEW

**Comment by zowoq** [comment]
```diff
diff --git a/pkgs/applications/networking/cluster/k3s/update.sh b/pkgs/applications/networking/cluster/k3s/update.sh
index b9d774847d3..0231fd81ae9 100755
--- a/pkgs/applications/networking/cluster/k3s/update.sh
+++ b/pkgs/applications/networking/cluster/k3s/update.sh
@@ -65,7 +65,6 @@ setKV k3sVersion ${K3S_VERSION}
 setKV k3sCommit ${K3S_COMMIT}
 setKV k3sRepoSha256 ${K3S_REPO_SHA256}
 
-setKV k3sVendorSha256 "0000000000000000000000000000000000000000000000000000"
 setKV k3sServerVendorSha256 "0000000000000000000000000000000000000000000000000000"
 
 setKV traefikChartVersion ${TRAEFIK_CHART_VERSION}
@@ -100,14 +99,7 @@ fi
 #         And then collect the vendorSha256 from that. But as this is an internal package.
 #         It wouldn't be pretty. Any other solution?
 #
-# set +e
-# K3S_VENDOR_SHA256=$(nix-build --no-out-link -A k3s 2>&1 >/dev/null | grep "got:" | cut -d':' -f2 | sed 's| ||g')
-# set -e
-# cd - > /dev/null
 
-# if [ -n "${K3S_VENDOR_SHA256:-}" ]; then
-#   setKV k3sVendorSha256 ${K3S_VENDOR_SHA256}
-# else
-#   echo "Update failed. K3S_VENDOR_SHA256 is empty."
-#   exit 1
-# fi
+K3S_VENDOR_SHA256=$(nix-prefetch -I nixpkgs=./. "{ sha256 }: (import ./. {}).k3s.go-modules.overrideAttrs (_: { vendorSha256 = sha256; })")
+
+setKV k3sVendorSha256 ${K3S_VENDOR_SHA256}

```
Untested example but should work as the vendor hash from main package. `nixpkgs=./.` and `import ./.` do need to be adjusted relative to the script directory.

`k3sServerVendorSha256` is a bit harder to do as an internal package though, not sure offhand of the right way use the same method for it.

**Comment by superherointj** [comment]
@zowoq Thanks for the helpful answer. The script seems to be working fine now. But we can only be sure when it is run by the r-bot.

I ask reviewers to review & merge this.

**Review by euank — COMMENTED** [review]
This overall looks good to me!

We could split out the `k3sServer` derivation into its own file if we wanted, which could allow using `nix-prefetch` for it too.

The whole `setKV` dance still seems a bit hacky compared to, say, generating a standalone file we could import with all the bits that change... but I guess there's maybe some other tradeoffs there too.

**Comment by superherointj** [comment]
> This overall looks good to me!

:)

> We could split out the `k3sServer` derivation into its own file if we wanted, which could allow using `nix-prefetch` for it too.
> 
> The whole `setKV` dance still seems a bit hacky compared to, say, generating a standalone file we could import with all the bits that change... but I guess there's maybe some other tradeoffs there too.

A standalone file would be useful for offering parallel versions of K3s as well. Which I have missed myself when trying to temporarily avoid a latest Kubernetes change.

Can we merge this as it is and later we put more thought into this?

File: pkgs/applications/networking/cluster/k3s/update.sh [hunk 1]
```diff
@@ -62,3 +73,36 @@ setKV k3sRootSha256 ${K3S_ROOT_SHA256}
...
 setKV k3sCNISha256 ${CNIPLUGINS_SHA256}
+
+setKV containerdVersion ${CONTAINERD_VERSION}
+setKV containerdSha256 ${CONTAINERD_SHA256}
+
+setKV criCtlVersion ${CRI_CTL_VERSION}
+
+setKV k3sServerVendorSha256 "0000000000000000000000000000000000000000000000000000"
+
+cd ../../../../../
```
[the remark this answers is not in the document]
Comment (superherointj) [reply pkgs/applications/networking/cluster/k3s/update.sh]: Done

**Review by euank — APPROVED** [review]
Merging this and seeing if it starts the auto-updates flowing sgtm :+1: 

We can make further improvements as additional PRs if we'd like

**Review by jonringer — APPROVED** [review]

**Review by jonringer — APPROVED** [review]
LGTM

Result of nixpkgs-review pr 161906 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>1 package built:</summary>  <ul>    <li>k3s</li>  </ul></details>

**Pull request merged**